pkg/kv/kvserver/batcheval/cmd_end_transaction.go

```diff
@@ -1217,6 +1232,30 @@ func mergeTrigger(
 	pd.Replicated.Merge = &kvserverpb.Merge{
 		MergeTrigger: *merge,
 	}
+
+	{
+		// If we have GC hints populated that means we are trying to perform
+		// optimized garbage removal in future.
+		// We will try to merge both hints if possible and set new hint on LHS.
+		lhsLoader := MakeStateLoader(rec)
+		lhsHint, err := lhsLoader.LoadGCHint(ctx, batch)
+		if err != nil {
+			return result.Result{}, err
+		}
+		rhsLoader := stateloader.Make(merge.RightDesc.RangeID)
+		rhsHint, err := rhsLoader.LoadGCHint(ctx, batch)
+		if err != nil {
+			return result.Result{}, err
+		}
+		if lhsHint.Merge(rhsHint) {
```

Comment:
I *think* this is correct but also it's brittle? If the cluster setting hasn't been bumped yet, we'll get `false` here, because the right-hand side can't possibly have a GCHint, and so we're definitely not getting a `true` here and thus are not calling `SetGCHint`.

Looking at the way all the plumbing works now I am a little confused why we don't put the bool on `SetGCHint`. By definition, anyone writing the GCHint has to go through that method. Yet, at the moment that method will happily write a GCHint even when it's not allowed and we have to check each caller.

If the method instead took a `GCHintsEnabled bool`, every caller would have to be aware of it.

I just pushed a commit that attempts it to this PR. Mind taking a look and seeing if you prefer it? (I do).